### PR TITLE
Fix parsing of IPv4 addresses

### PR DESCRIPTION
**Pull request by garyb**
Also adds some test infrastructure for property tests. Just tests the IPv4 stuff for now, but will be useful in the future I think.

### CODE REVIEW

File: src/Data/URI/Host/Gen.purs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+module Data.URI.Host.Gen where
+
+import Prelude
+
+import Control.Monad.Gen as Gen
+import Data.String as S
+import Data.URI.Host (Host(..))
+
+genIPv4 :: forall m. Gen.MonadGen m => m Host
+genIPv4 = do
```
Comment (kritzcreek) [inline src/Data/URI/Host/Gen.purs:10]: This will not generate ips like `123.001.123.123` though, because of the `001` in there.

**Review by kritzcreek — COMMENTED** [review]
Looks good apart from my comment!

File: src/Data/URI/Host/Gen.purs (hunk 1 above)
Comment (garyb) [reply src/Data/URI/Host/Gen.purs:10]: Is that a valid notation for IPv4 addresses? I haven't seen them formatted that way before.

I think the parser will work for that actually, so should either improve the generator or make the parser stricter I guess!

Comment (kritzcreek) [reply src/Data/URI/Host/Gen.purs:10]: I mean they actually just correspond to one byte each. I don't think the decimal notation has a formal specification anywhere, but googling around yields that it's preferred to not allow prefixed 0's as they can lead some number parsers to assume the octal system.

Comment (garyb) [reply src/Data/URI/Host/Gen.purs:10]: Here's the bit from the rfc3986 spec actually, looks like leading zeros shouldn't be accepted:
```
      IPv4address = dec-octet "." dec-octet "." dec-octet "." dec-octet

      dec-octet   = DIGIT                 ; 0-9
                  / %x31-39 DIGIT         ; 10-99
                  / "1" 2DIGIT            ; 100-199
                  / "2" %x30-34 DIGIT     ; 200-249
                  / "25" %x30-35          ; 250-255
```

**Review by kritzcreek — APPROVED** [review]

**Pull request merged**